lib/pf/web/dispatcher.pm

```diff
@@ -53,205 +52,184 @@ Reference: http://perl.apache.org/docs/2.0/user/handlers/http.html#PerlTransHand
+    }
+        
+    # Enforcing HTTP (not HTTPS) when dealing with captive-portal detection mecanism
+    if ( ($url =~ /$WEB::CAPTIVE_PORTAL_DETECTION_MECANISM_URLS/o) || ($user_agent =~ /CaptiveNetworkSupport/s) ) {
+        $proto = $HTTP;
+        $logger->info("We are dealing with a device with captive portal detection capabilities. " .
+            "We are using HTTP rather than HTTPS to avoid SSL certificate related errors");
+    }
 
-     my $captiv_url = APR::URI->parse($r->pool,"$proto://".$r->hostname."/captive-portal");
-    $captiv_url->query($r->args);
+    # Configuring redirect URLs for both the portal and the WISPr(need to be part of the header in case of a WISPr client)
+    my $portal_url = APR::URI->parse($r->pool,"$proto://".${captive_portal_domain}."/captive-portal");
+    $portal_url->query("destination_url=$destination_url&".$r->args);
+    my $wispr_url = APR::URI->parse($r->pool,"$proto://".${captive_portal_domain}."/wispr");
+    $wispr_url->query($r->args);
+
+    # External captive-portal / Webauth handling
+    # In the case of an external captive-portal, we want to use a different URL (the hostname to which the network equipment send the request, which is PacketFence but maybe not the configured hostname in pf.conf)
+    # We also need to keep track of the CGI session by setting a cookie
+    my $external_portal = pf::web::externalportal->new;
+    my ( $cgi_session_id, $external_portal_destination_url ) = $external_portal->handle($r);
+    if ( $cgi_session_id ) {
+        $logger->debug("We are dealing with an external captive-portal / webauth request. Adjusting the redirect URL accordingly");
+        $r->err_headers_out->add('Set-Cookie' => "CGISESSION_PF=".  $cgi_session_id . "; path=/");
+        $destination_url = $external_portal_destination_url if ( defined($external_portal_destination_url) );
+
+        # Re-Configuring redirect URLs for both the portal and the WISPr(need to be part of the header in case of a WISPr client)
+        $portal_url = APR::URI->parse($r->pool,"$proto://".$r->hostname."/captive-portal");
+        $portal_url->query("destination_url=$destination_url&".$r->args);
+        $wispr_url = APR::URI->parse($r->pool,"$proto://".$r->hostname."/wispr");
+        $wispr_url->query($r->args);
+    }
 
     my $stash = {
-        'login_url' => $captiv_url->unparse(),
+        'portal_url' => $portal_url->unparse(),,
+        'wispr_url' => $wispr_url->unparse(),,
     };
 
-
-
-
-    # prepare custom REDIRECT response
     my $response = '';
     my $template = Template->new({
         INCLUDE_PATH => [$CAPTIVE_PORTAL{'TEMPLATE_DIR'}],
     });
-    $template->process( "redirection.html", $stash, \$response ) || $logger->error($template->error());;
+    $template->process("redirect.tt", $stash, \$response) || $logger->error($template->error());
+
+    $r->headers_out->set('Location' => $stash->{portal_url});
     $r->content_type('text/html');
     $r->no_cache(1);
```

Comment:
This doesn't work, let's change with:
$r->err_headers_out->add('Cache-Control' => "no-cache, no-store, must-revalidate");
$r->err_headers_out->add('Pragma' => "no-cache");
$r->err_headers_out->add('Expire' => '10');